### PR TITLE
OPERATOR-752: cleanup px created configmaps

### PR DESCRIPTION
**Pull request by haibinxie**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** (optional)
Closes #

**Special notes for your reviewer**:

### CODE REVIEW

**Review by piyush-nimbalkar — COMMENTED** [review]
Not sure if we need to do all of this to just delete the config maps. We already have some logic to delete configmaps - https://github.com/libopenstorage/operator/blob/master/drivers/storage/portworx/uninstall.go#L133
We can just change that code to list and delete the known configmaps.

**Comment by haibinxie** [comment]
> Not sure if we need to do all of this to just delete the config maps. We already have some logic to delete configmaps - https://github.com/libopenstorage/operator/blob/master/drivers/storage/portworx/uninstall.go#L133 We can just change that code to list and delete the known configmaps.

@piyush-nimbalkar i moved this to operator bcz this is general, after this PR we won't need to handle deletion in the future anymore, anyone (px, telemetry) wants to let operator delete something just add the annotation in their object.

**Comment by piyush-nimbalkar** [comment]
> i moved this to operator bcz this is general, after this PR we won't need to handle deletion in the future anymore, anyone (px, telemetry) wants to let operator delete something just add the annotation in their object.

Sounds good to me and will work well in this situation. 
In general and for future, we should be careful when we add operator awareness in px. It could cause dependency from both sides and make changing operator code very difficult to maintain backward compatibility.

File: pkg/controller/storagecluster/storagecluster.go [hunk 1]
```diff
@@ -649,6 +658,45 @@ func (c *Controller) deleteStorageCluster(
...
+	if obj.GetObjectKind().GroupVersionKind().Kind == "ConfigMap" {
+		if obj.GetName() == "px-attach-driveset-lock" ||
+			strings.HasPrefix(obj.GetName(), "px-bringup-queue-lock") {
+			return true
+		}
+	}
+
+	return false
+}
+
```
Comment (piyush-nimbalkar) [inline pkg/controller/storagecluster/storagecluster.go:699]: This is very portworx specific. Can we move at least this or the entire removeResources function to the portworx driver? Let's not spill px specific code in the controller.

File: pkg/util/k8s/k8s.go [hunk 2]
```diff
@@ -1836,3 +1836,173 @@ func GetV1beta1CRDFromFile(
 	}
 	return crd, nil
 }
+
+// GetAllObjects gets all objects from k8s
+func GetAllObjects(k8sClient client.Client, namespace string) ([]client.Object, error) {
```
Comment (piyush-nimbalkar) [inline pkg/util/k8s/k8s.go:1841]: nit: Do we really need to go overboard and include all the objects?
I guess just secrets, configmaps, services, endpoints and storageclasses should do it.

File: pkg/util/k8s/k8s.go [hunk 3]
```diff
@@ -1836,3 +1836,173 @@ func GetV1beta1CRDFromFile(
...
+	} else if l, ok := list.(*monitoringv1.PrometheusList); ok {
+		for _, o := range l.Items {
+			o.Kind = "Prometheus"
+			*objs = append(*objs, o.DeepCopy())
+		}
+	} else if l, ok := list.(*monitoringv1.AlertmanagerList); ok {
+		for _, o := range l.Items {
+			o.Kind = "Alertmanager"
+			*objs = append(*objs, o.DeepCopy())
+		}
```
Comment (piyush-nimbalkar) [inline pkg/util/k8s/k8s.go:2000]: Do we need to check all these objects?

File: pkg/controller/storagecluster/storagecluster.go [hunk 4]
```diff
@@ -649,6 +658,45 @@ func (c *Controller) deleteStorageCluster(
...
+	for _, obj := range objs {
+		gcAnnotated := false
+		if obj.GetAnnotations() != nil {
+			gcAnnotated, _ = strconv.ParseBool(obj.GetAnnotations()[constants.AnnotationGarbageCollection])
+		}
+
+		if gcAnnotated || c.gcNeeded(obj) {
+			logrus.Infof("Garbage collect object %s %s", obj.GetObjectKind().GroupVersionKind().Kind, obj.GetName())
+			if err := c.client.Delete(context.TODO(), obj); err != nil {
```
Comment (pure-jliao) [inline pkg/controller/storagecluster/storagecluster.go:675]: need to ignore `NotFound` errors

File: pkg/util/k8s/k8s.go (hunk 3 above)
Comment (pure-jliao) [reply pkg/util/k8s/k8s.go:2000]: doesn't hurt IMO probably need those somedays.

**Review by pure-jliao — COMMENTED** [review]
So moving forward resources added by operator would be managed by operator of course, external added resources like other config maps generated by Portworx, the creator should be responsible to add this annotation to the object right?

File: pkg/controller/storagecluster/controller_test.go [hunk 5]
```diff
@@ -2736,6 +2736,154 @@ func TestUpdateDriverWithInstanceInformation(t *testing.T) {
...
 }
 
+func TestGarbageCollection(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	cluster := createStorageCluster()
+
+	// Configmaps created by px are expected to be deleted https://portworx.atlassian.net/browse/OPERATOR-752
+	configMaps := []v1.ConfigMap{
```
Comment (dsupure) [inline pkg/controller/storagecluster/controller_test.go:2746]: Can we have a list and append 1 by 1? Instead hard-coded inside of the test?

File: pkg/util/k8s/k8s.go [hunk 6]
```diff
@@ -1836,3 +1836,173 @@ func GetV1beta1CRDFromFile(
 	}
 	return crd, nil
 }
+
+// GetAllObjects gets all objects from k8s
+func GetAllObjects(k8sClient client.Client, namespace string) ([]client.Object, error) {
+	var objs []client.Object
+
+	if err := AppendObjectList(k8sClient, namespace, &v1.ServiceList{}, &objs); err != nil {
```
Comment (dsupure) [inline pkg/util/k8s/k8s.go:1844]: Can we change this to be a loop?

File: pkg/util/k8s/k8s.go [hunk 7]
```diff
@@ -1836,3 +1836,173 @@ func GetV1beta1CRDFromFile(
...
+		list,
+		&client.ListOptions{
+			Namespace: namespace,
+		})
+	if err != nil {
+		logrus.WithError(err).Errorf("failed to list objects %v", list.GetObjectKind())
+		return err
+	}
+
+	if l, ok := list.(*v1.ServiceList); ok {
```
Comment (dsupure) [inline pkg/util/k8s/k8s.go:1906]: Can we change this to be a loop?

File: pkg/controller/storagecluster/storagecluster.go (hunk 4 above)
Comment (haibinxie) [reply pkg/controller/storagecluster/storagecluster.go:675]: It should not happen as the object is found before deleting it.

File: pkg/util/k8s/k8s.go (hunk 6 above)
Comment (haibinxie) [reply pkg/util/k8s/k8s.go:1844]: good idea, i will make a change

File: pkg/util/k8s/k8s.go (hunk 7 above)
Comment (haibinxie) [reply pkg/util/k8s/k8s.go:1883]: looks like this type conversion is hard to do in list, but i changed the other one into loop

File: pkg/controller/storagecluster/controller_test.go (hunk 5 above)
Comment (dsupure) [reply pkg/controller/storagecluster/controller_test.go:2746]: I mean do a for loop.

**Review by dsupure — APPROVED** [review]
LGTM!

**Review by pure-jliao — APPROVED** [review]

**Pull request merged**